### ISSUE TITLE
Dependency override causes builders to be applied at the top level incorrectly

### ISSUE THREAD

**Issue by alanknight-wk** [dart-lang/build#3009]
* Dart SDK Version (`dart --version`)
 * Dart SDK version: 2.12.0-133.7.beta (beta) (Tue Jan 12 09:25:38 2021 +0100) on "macos_x64"
 * also 2.7.2 for comparison
* What package(s) from this repo you are using, and the version (i.e. `build_runner 0.7.12`)
  * _If you are not sure, try checking your `pubspec.lock` file._
  * build_runner 1.11.1
  * build_resolvers 1.5.3
  * build_modules 3.0.4
* What builder(s) you are using (or writing yourself). Try to give a short summary of what they do.
  * I was testing with built_redux, but built_value_generator exhibits the same behavior
* Whether you are using Windows, MacOSX, or Linux (if applicable)
  * Mac

TL;DR.  In 2.7.2 adding a dependency override of a builder to a package that did not directly depend on the builder did not change the behavior. In 2.12 it seems to incorrectly apply the builder to the top-level package.

I discussed this with @jakemac53 in the channel. In looking at upgrading to 2.12 we saw a significant performance regression. This turned out to be a lot of time spent in builders. And removing one builder just pushed the time to a different one.

Because we were testing upgrading we had to tweak a number of packages, both ours and others, and we added dependency overrides to our main package. The main package did not use either built_value_generator or built_redux, but had dependencies that did. The build times became significantly slower. I could reproduce this in 2.7.2, but only if I added built_redux as a direct dependency rather than as an override.

So I think the correct behavior with a dependency override is
  - for any package that depended on package_with_builder, apply the builder, but using the overridden version
  - for any package that did not normally depend on package_with_builder, do not apply the builder

That seems to be different between 2.7.2 and 2.12, though probably not in the SDK but rather in one of the packages. I notice that package_config.json doesn't seem to have the distinction between a direct and a transitive dependency, and I wonder if maybe that's involved - maybe if something started getting information from package_config.json instead of pubspec.lock it might have lost that distinction and treated everything like a direct dependency for builder-applying purposes.

**Comment by alanknight-wk** [comment]
Also, the reason that the time would transfer from one builder to another is that the main cost was the analysis of all the package files. So running N builders didn't cost significantly more than running one.

**Comment by jakemac53** [comment]
I believe we just need to remove this line https://github.com/dart-lang/build/blob/master/build_runner_core/lib/src/package_graph/package_graph.dart#L219.

cc @natebosch should we consider that breaking do you think?

**Comment by jakemac53** [comment]
@alan-knight can you try using a git override to depend on https://github.com/dart-lang/build/pull/3011 and see if that fixes the issue for you?

**Comment by natebosch** [comment]
> cc @natebosch should we consider that breaking do you think? I lean towards not.

I think we should consider it non-breaking. That behavior was there [since the beginning](https://github.com/dart-lang/build/pull/11/files). The reason it was written then was because we wanted to override the dependency type and we were collecting both names and sources in the same method. `dev_dependencies` were added in #63

I think what should have happened is that in #1306 I should have stopped reading overrides altogether since we stopped using the pubspec as the way to find the dependency types. Prior to that I think the bug still existed because we should have only assigned with the overrides if they were already in the map, so my PR didn't change behavior but did make the bug more obvious.
